### DISCUSSION TITLE
Add changing lights as an alternative to a physical robot when GLaDOS is speaking

### DISCUSSION THREAD

**Discussion by salleq** [nerdaxic/glados-voice-assistant#12] [Show and tell]
This could be defined in the settings.env (whether using servos etc. or home assistant)
I'm thinking about commanding a RGB led strip (but it could be anything) which would:

- Change color whenever GLaDOS starts to speak
- Change color again whenever GLaDOS stops speaking

So basically there could be defined either:

1.  two different scenes
2.  two different service calls to Home Asisstant
3. or a mix of these

**Comment by nerdaxic** [comment]
I like the idea, it is already half-way implemented already for the ReSpeaker LED-ring and the single NeoPixel "rec light" next to the eye.
Could simply add hooks to Home Assistant. 

Same hooks could be used to for example used for things like "mute spotify when GLaDOS is listening" and then return it the original volume.

Added triggers like "start_listening" into the config file suggestion. [Would this solve the use-case? ](https://github.com/nerdaxic/glados-voice-assistant/discussions/5)

**Comment by salleq** [comment]
Yes, this would definitely solve it the need, as long as there's freedom to call whatever service in HA (could be something else than a script). But of course if this is difficult then just the script will suffice.

**Reply by nerdaxic** [reply]
I shall start with the script and keep that in mind.
This goes to todo pile.

**Comment by salleq** [comment]
Should there also be a different hook for erorrs?
Right now GLaDOS speaking out errors (STT error, STT API error, didn't understand what you said) uses the same speak-function as the normal speak functions, so say if you wanted to have red light when there's an error and green for normal speech, there would have to be three different functions: started normal speaking, started error speaking, stopped speaking

And maybe change the function of those errors from speak() to this third function.

Definitely not a priority, but a thought again.

EDIT: of course - in addition to started_listening and stopped_listening.

So this would be 5 total hooks:

- started listening
- stopped listening
- started speaking
- started (error speaking)
- stopped speaking

**Comment by salleq** [comment]
This is how I got this working by changing glados_functions.py

I'm definitely not certain this is the best way to do it, if at all a reasonable way to do it. That's why I didn't open a PR yet.
If I try to import the runHaScript function from gladosHA.py, I keep getting errors on *speak* not being defined whenever sayNumericSensorData is called. Doesn't seem to help even if the whole gladosHA is imported (again).  I'm a newbie with Python altogether but there's either something that I don't get on importing functions OR it's safer to redefine the few lines again.


```python
#!/usr/bin/python3
from glados_respeaker import *
from gladosSerial import *
from gladosServo import *
from word2number import w2n
from subprocess import call
import re
import os

import requests 
#import gladosHA as HA
#from gladosHA import runHaScript
from dotenv import load_dotenv
load_dotenv(dotenv_path=os.path.dirname(os.path.abspath(__file__))+'/settings.env')

token = os.getenv('HOME_ASSISTANT_TOKEN')
endpoint = os.getenv('HOME_ASSISTANT_ADDRESS')+"/api/"

script_startListening = os.getenv('HOME_ASSISTANT_SCRIPT_STARTLISTENING')
script_stopListening = os.getenv('HOME_ASSISTANT_SCRIPT_STOPLISTENING')
script_startSpeaking = os.getenv('HOME_ASSISTANT_SCRIPT_STARTSPEAKING')
script_stopSpeaking = os.getenv('HOME_ASSISTANT_SCRIPT_STOPSPEAKING')
script_errorSpeaking = os.getenv('HOME_ASSISTANT_SCRIPT_ERRORSPEAK') # TODO: separate error speaking from normal speaking

def runHaScript(script):
	url = endpoint+"services/script/"+script
	headers = {
	 "Authorization": "Bearer "+token,
	 "content-type": "application/json",
	}

	response = requests.post(url, headers=headers, verify=False)

def started_listening():
	print("started listening")
	runHaScript(script_startListening)
	if(os.getenv('RESPEAKER_CONNECTED')):
		respeaker_mode("listen")
```
...

**Reply by nerdaxic** [reply]
Thanks for sharing!
This functionality will come with home assistant yaml config, so I won't be adding it to the settings file for now.

**Reply by salleq** [reply]
Yeah of course! I got so carried away with tinkering with it that I forgot they were supposed to be moved to YAML 😂

**Comment by nerdaxic** [comment]
Implemented. Scripts can be set up in the [home_assistant_settings.yaml](https://github.com/nerdaxic/glados-voice-assistant/blob/main/settings/home_assistant_settings.yaml.sample) file.
